modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1076 @@
+  // To access them from another node, make sure you use 'getPrivatizedThis'
+  pragma "no doc"
+  var bag : Bag(eltType);
+
+  proc DistributedBag(type eltType, targetLocales : [?targetLocDom] locale = Locales) {
+    bag = new Bag(eltType, this);
+    this.targetLocDom = targetLocDom;
+    this.targetLocales = targetLocales;
+    pid = _newPrivatizedClass(this);
+  }
+
+  pragma "no doc"
+  proc DistributedBag(other, pid, type eltType = other.eltType) {
+    bag = new Bag(eltType, this);
+    this.targetLocDom = other.targetLocDom;
+    this.targetLocales = other.targetLocales;
+    this.pid = pid;
+  }
+
+  pragma "no doc"
+  proc dsiPrivatize(pid) {
+    return new DistributedBag(this, pid);
+  }
+
+  pragma "no doc"
+  proc dsiGetPrivatizeData() {
+    return pid;
+  }
+
+  pragma "no doc"
+  inline proc getPrivatizedThis {
+    if this.locale == here {
+      return this;
+    }
+
+    return chpl_getPrivatizedCopy(this.type, pid);
+  }
+
+  /*
+    Obtains a privatized version of this instance. The privatized version is a
+    cloned instance that is allocated on this node, which is useful for eliding
+    communications generated from accessing instance fields. Using another node's
+    instance will significantly penalize performance by bounding overall throughput
+    on the communication between the two nodes.
+  */
+  proc getPrivatizedInstance() {
+    return getPrivatizedThis;
+  }
+
+  /*
+    Insert an element to this node's bag. The ordering is not guaranteed to be
+    preserved.
+  */
+  proc add(elt : eltType) : bool {
+    return getPrivatizedThis.bag.add(elt);
+  }
+
+  /*
+    Remove an element from this node's bag. The order in which elements are removed
+    are not guaranteed to be the same order it has been inserted. If this node's
+    bag is empty, it will attempt to steal elements from bags of other nodes.
+  */
+  proc remove() : (bool, eltType) {
+    return getPrivatizedThis.bag.remove();
+  }
+
+  /*
+    Obtain the number of elements held in all bags across all nodes. This method
+    is best-effort and can be non-deterministic for concurrent updates across nodes,
+    and may miss elements or even count duplicates resulting from any concurrent
+    insertion or removal operations.
+  */
+  proc getSize() : int {
+    var sz : atomic int;
+    forall loc in targetLocales do on loc {
+      var instance = getPrivatizedThis;
+      forall segmentIdx in 0..#here.maxTaskPar {
+        sz.add(instance.bag.segments[segmentIdx].nElems.read() : int);
+      }
+    }
+
+    return sz.read();
+  }
+
+  /*
+    Performs a lookup to determine if the requested element exists in this bag.
+    This method is best-effort and can be non-deterministic for concurrent
+    updates across nodes, and may miss elements resulting from any concurrent
+    insertion or removal operations.
+  */
+  proc contains(elt : eltType) : bool {
+    var foundElt : atomic bool;
+    forall elem in getPrivatizedThis {

```

Comment:
I havn't dug in to the these call below, but I would've expected this one to be written with somehting like
``` chapel
coforall loc in targetLocales do on loc {
  forall elem in getPrivatizedThis.localElements() {
        if elem == elt {
          foundElt.write(true);
       }
  }
}
```